pkg/reporter/reporter.go

```diff
@@ -6,72 +6,164 @@ import (
 
 	reqBody, err := json.Marshal(report)
 	if err != nil {
-		return fmt.Errorf("failed to serialize results as JSON: %w", err)
+		return nil, fmt.Errorf("failed to serialize results as JSON: %w", err)
 	}
 
-	req, err := http.NewRequest("POST", anchoreURL, bytes.NewBuffer(reqBody))
+	req, err := http.NewRequest("POST", apiEndpoint, bytes.NewBuffer(reqBody))
 	if err != nil {
-		return fmt.Errorf("failed to build request to report data to Anchore: %w", err)
+		return nil, fmt.Errorf("failed to build request to report data to Anchore: %w", err)
 	}
 	req.SetBasicAuth(anchoreDetails.User, anchoreDetails.Password)
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("x-anchore-account", anchoreDetails.Account)
-	resp, err := client.Do(req)
+
+	return req, nil
+}
+
+type AnchoreVersion struct {
+	API struct {
+		Version string `json:"version"`
+	} `json:"api"`
+	DB struct {
+		SchemaVersion string `json:"schema_version"`
+	} `json:"db"`
+	Service struct {
+		Version string `json:"version"`
+	} `json:"service"`
+}
+
+func fetchVersionedAPIPath(anchoreDetails connection.AnchoreInfo) (string, error) {
+	logger.Log.Debug("Detecting Anchore API version")
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: anchoreDetails.HTTP.Insecure},
+	} // #nosec G402
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   time.Duration(anchoreDetails.HTTP.TimeoutSeconds) * time.Second,
+	}
+	gock.InterceptClient(client) // Required to use gock for testing custom client
+
+	versionEndpoint, err := url.JoinPath(anchoreDetails.URL, "version")
 	if err != nil {
-		if resp != nil {
-			if resp.StatusCode == 401 {
-				return fmt.Errorf("failed to report data to Anchore, check credentials: %w", err)
-			}
-		}
-		return fmt.Errorf("failed to report data to Anchore: %w", err)
+		return v1ReportAPIPath, fmt.Errorf("failed to parse API URL: %w", err)
+	}
+
+	resp, err := client.Get(versionEndpoint)
+	if err != nil {
+		return v1ReportAPIPath, fmt.Errorf("failed to contact Anchore API: %w", err)
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode < 200 || resp.StatusCode > 299 {
-		return fmt.Errorf("failed to report data to Anchore: %+v", resp)
+	if resp.StatusCode != 200 {
+		return v1ReportAPIPath, fmt.Errorf("failed to retrieve Anchore API version: %+v", resp)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return v1ReportAPIPath, fmt.Errorf("failed to read Anchore API version: %w", err)
 	}
-	logger.Log.Debug("Successfully reported results to Anchore", "Account", anchoreDetails.Account)
-	return nil
-}
 
-func buildURL(anchoreDetails connection.AnchoreInfo) (string, error) {
-	anchoreURL, err := url.Parse(anchoreDetails.URL)
+	ver := AnchoreVersion{}
+	err = json.Unmarshal(body, &ver)
 	if err != nil {
-		return "", err
+		return v1ReportAPIPath, fmt.Errorf("failed to parse API version: %w", err)
 	}
 
-	anchoreURL.Path += ReportAPIPath
+	logger.Log.Debug("Anchore API version: ", ver)
```

Comment:
This results in an error against Enterprise 4.8:

```
2023-08-21T11:44:19.766+0100    debug   logger/logger.go:31     Detecting Anchore API version
2023-08-21T11:44:19.768+0100    error   zap@v1.24.0/sugar.go:204        Ignored key without a value.    {"ignored": {"api":{"version":""},"db":{"schema_version":"24"},"service":{"version":"4.8.1"}}}
go.uber.org/zap.(*SugaredLogger).Debugw
        /Users/bradley/go/pkg/mod/go.uber.org/zap@v1.24.0/sugar.go:204
github.com/anchore/ecs-inventory/internal/logger.ZapLogger.Debug
        /Users/bradley/src/ecs-inventory/internal/logger/logger.go:31
github.com/anchore/ecs-inventory/pkg/reporter.fetchVersionedAPIPath
        /Users/bradley/src/ecs-inventory/pkg/reporter/reporter.go:162
github.com/anchore/ecs-inventory/pkg/reporter.Post
        /Users/bradley/src/ecs-inventory/pkg/reporter/reporter.go:59
github.com/anchore/ecs-inventory/cmd.glob..func2
        /Users/bradley/src/ecs-inventory/cmd/root.go:52
github.com/spf13/cobra.(*Command).execute
        /Users/bradley/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920
github.com/spf13/cobra.(*Command).ExecuteC
        /Users/bradley/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044
github.com/spf13/cobra.(*Command).Execute
        /Users/bradley/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
github.com/anchore/ecs-inventory/cmd.Execute
        /Users/bradley/src/ecs-inventory/cmd/cmd.go:38
main.main
        /Users/bradley/src/ecs-inventory/main.go:6
runtime.main
        /opt/homebrew/opt/go@1.20/libexec/src/runtime/proc.go:267
        
```

Reply:
^^ Above suggestion will work it just prints out the object in the log rather than trying to do the structured logging of the object. I think this is fine as it's just a debug log. k8s-inventory didn't hit this because it hasn't moved to zap yet

Reply:
With the zap structure should it be this?

```
logger.Log.Debug("Anchore API version: ", "version", ver)
```

Reply:
Actually there's no point doing that really, may as well just print it in the string as you said

Reply:
Yeah if that works then that's fine too